File: LLM-Translations/GPT-4/codenet/Go/Java/s269502139.java

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

class Node {
    int key;
    Node prev;
    Node next;

    Node(int key) {
        this.key = key;
    }

    void rPrint(StringBuilder buf) {
        buf.append(key);
        if (next != null) {
            buf.append(" ");
            next.rPrint(buf);
        }
    }

    Node[] delete(Node first, Node last) {
        if (prev != null && next != null) {
            prev.next = next;
            next.prev = prev;
            return new Node[]{first, last};
        }

        if (next == null && prev == null) {
            return new Node[]{null, null};
        }

        if (prev == null) {
            first = next;
            next.prev = null;
        }

        if (next == null) {
            last = prev;
            prev.next = null;
        }

        return new Node[]{first, last};
    }
}

public class s269502139 {
    public static void main(String[] args) throws IOException {
        BufferedReader stdin = new BufferedReader(new InputStreamReader(System.in));
        Node first = null, last = null;
        stdin.readLine();

        String cmd;
        while ((cmd = stdin.readLine()) != null) {
            String[] xcmd = cmd.split(" ");
            switch (xcmd[0]) {
                case "deleteFirst":
                    Node[] firstDelete = first.delete(first, last);
                    first = firstDelete[0];
                    last = firstDelete[1];
                    break;
                case "deleteLast":
                    Node[] lastDelete = last.delete(first, last);
                    first = lastDelete[0];
                    last = lastDelete[1];
                    break;
                default:
                    int k = Integer.parseInt(xcmd[1]);
                    switch (xcmd[0]) {
                        case "delete":
                            for (Node a = first; a != null; a = a.next) {
                                if (a.key == k) {
                                    Node[] deleteNode = a.delete(first, last);
                                    first = deleteNode[0];
                                    last = deleteNode[1];
                                    break;
                                }
                            }
                            break;
                        case "insert":
                            Node n = new Node(k);
                            n.next = first;
                            if (first != null) {
                                first.prev = n;
                            }

                            first = n;
                            if (last == null) {
                                last = n;
                            }
                            break;
                    }
                    break;
            }
        }
        StringBuilder buf = new StringBuilder();
        first.rPrint(buf);
        System.out.println(buf.toString());
    }
}

// 

File: LLM-Translations/GPT-4/codenet/Go/Python/s269502139.py

class Node:
    def __init__(self, key):
        self.key = key
        self.prev = None
        self.next = None

    def r_print(self):
        result = []
        current = self
        while current is not None:
            result.append(str(current.key))
            current = current.next
        return " ".join(result)

    def delete(self, first, last):
        if self.prev is not None and self.next is not None:
            self.prev.next = self.next
            self.next.prev = self.prev
            return first, last

        if self.next is None and self.prev is None:
            return None, None

        if self.prev is None:
            first = self.next
            self.next.prev = None

        if self.next is None:
            last = self.prev
            self.prev.next = None

        return first, last

def main():
    n = int(input())
    first, last = None, None

    for _ in range(n):
        cmd = input().split()
        if cmd[0] == "deleteFirst":
            first, last = first.delete(first, last)
        elif cmd[0] == "deleteLast":
            first, last = last.delete(first, last)
        else:
            k = int(cmd[1])
            if cmd[0] == "delete":
                current = first
                while current is not None:
                    if current.key == k:
                        first, last = current.delete(first, last)
                        break
                    current = current.next
            elif cmd[0] == "insert":
                new_node = Node(k)
                new_node.next = first
                if first is not None:
                    first.prev = new_node
                first = new_node
                if last is None:
                    last = new_node

    print(first.r_print())

if __name__ == "__main__":
    main()



